Move file consider destinations which are prefix of the source

Arthur-Milchior opened the pull request:
I added a method to detect prefix of files.

As I found the code quite confusing to read, I rewrote it to split in three cases:
* source does not exists
* destination exists
* move can occur as expected

Fixed: #13170

david-allison commented:
Tests are failing

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/FileUtil.kt:
```diff
@@ -206,4 +208,45 @@ object FileUtil {
+                val fullFileContent = fullFileBuffer.read()
+                val prefixFileEnded = prefixContent == -1
+                val fullFileEnded = fullFileContent == -1
+                if (prefixFileEnded && fullFileEnded) {
```
Maybe I'm missing something (reviewing on my phone), but this doesn't seem to handle file contents.

If one file contains "a" and the other contains "bb", then it should not be considered to be a prefix

david-allison replied:
Ah! Definitely an issue reviewing on my phone. Thanks

Arthur-Milchior reviewed (commented):
I'll look at the failures

Arthur-Milchior commented:
Corrected.
Only difference, I believe it makes more sense to have a progress of 0 reported all of the time a move's source is a non existing field. 
I also would be all right with always having no progress reported.

Just, having a progress reported if and only if the destination does not exists didn't make much sense to me, so I changed an implementation. And so changed one test at the same time

david-allison reviewed (approved):
This is great, thank you!

david-allison commented:
nit: https://github.com/ankidroid/Anki-Android/blob/3f251cf6dc61e52ea9fb91116ab09fc17d84596b/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/migrateuserdata/MigrateUserData.kt#L535-L537 can be removed

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/MoveFile.kt:
```diff
@@ -39,56 +39,107 @@ import java.io.File
+            }
+            FileUtil.FilePrefix.STRICT_PREFIX -> {
+                val deletionSucceeded = destinationFile.delete()
+                Timber.w("(conflict) Deleted empty file in destination. Deletion result: %b", deletionSucceeded)
```
```suggestion
                Timber.w("(conflict) Deleted partial file in destination. Deletion result: %b", deletionSucceeded)
```

technically an empty file is a partial file isn't it? anyway now that you are checking for prefixes empty is definitely not correct, partial seems more correct. This is a nit though

mikehardy reviewed (approved):
All seems fine, I had a nit with a concrete suggestion, it is just changing wording a log message (which I attach very very low importance to)

Going to mark it approved, I'm fine with anyone merging this for speed without the nit fixed, or if the nit is fixed then *please* merge

mikehardy commented:
On reflection, if I'm going to hang this on a nit, I'm going to pull it locally, fix the nit then push and merge. No more delay
